File: CSharp/DceLib/Dce548.cs
namespace DceLib;

public static class Dce548
{
    /// <summary>
    /// Given a clock time in hours and minutes, get the angle between the
    /// hour hand and the minute hand, to the nearest degree
    /// </summary>
    /// <param name="time">the time, in hh:mm format</param>
    /// <returns>the angle between the hour hand and the minute hand, to the nearest degree</returns>
    public static int TimeToAngle(string time)
    {
        //each hour on the clock is 30 degrees
        int hour2deg = 30;
        
        //each minute is 6 degrees
        int min2deg = 6;

        //ensure time is 5 digits long
        if (time.Length != 5) throw new ArgumentException("argument too long");

        //split time into hours and minutes
        string[] times = time.Split(':');

        //ensure there is only 2 segments 
        if (times.Length != 2) throw new ArgumentException("argument must be split by a single ':'");

        //ensure that hours and minutes are only 2 digits
        if (times[0].Length != 2) throw new ArgumentException("hours and minutes should be 2 digits each");

        //try and parse hours into an int
        if (!int.TryParse(times[0], out int hours)) throw new ArgumentException("hours must be an integer");

        //ensure hours are valid
        if (hours > 23 || hours < 0) throw new ArgumentException("hours must be an integer from 00 to 23");
        if (hours > 12) hours -= 12;

        //try and parse minutes into an int
        if (!int.TryParse(times[1], out int minutes)) throw new ArgumentException("minutes must be an integer");

        //ensure minutes are valid
        if (minutes > 59 || minutes < 0) throw new ArgumentException("minutes must be an integer from 00 to 59");

        float thetaH = hours * hour2deg + minutes * hour2deg / 60.0f;
        float thetaM = minutes * min2deg;

        //convert hours and minutes to degrees, and then find the difference in the angles
        float result = MathF.Abs(thetaH - thetaM);

        //return the result
        return (int)MathF.Round(result);
    }
}

File: Python/Dce548/Dce548.py
class Dce548:
    @staticmethod
    def time_to_angle(time:str):

        hour2deg:int = 30
        
        min2deg:int = 6

        if time.__len__() != 5: raise ValueError("argument wrong size")

        times = time.split(":")

        if len(times) != 2: raise ValueError("argument must be split by a single ':'")

        if times[0].__len__() != 2: raise ValueError("hour and minutes must be 2 digits each")

        hours = int(times[0]) if times[0].isdecimal() else None

        if hours is None: raise ValueError("hour must be an integer between 00 and 23")

        if hours < 0 or hours > 23: raise ValueError("hour must be an integer between 00 and 23")
        if hours > 12: hours -= 12

        minutes = int(times[1]) if times[1].isdecimal() else None

        if minutes is None: raise ValueError("minutes must be an integer between 00 and 59")
        if minutes < 0 or minutes > 59: raise ValueError("minutes must be an integer between 00 and 59")

        theta_h:float = hours * hour2deg + minutes * hour2deg / 60
        theta_m:float = minutes * min2deg

        result:float = abs(theta_h - theta_m)

        return int(round(result))


